### PR TITLE
Add global context to store information about CPU environment

### PR DESCRIPTION
**Pull request by Vika-F**
- `cpu_info` class added to store the parameters of CPU; 
- `global_context` static class was added to store global information about the system.

### CODE REVIEW

File: cpp/oneapi/dal/detail/cpu.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
-* Copyright 2020 Intel Corporation
+* Copyright 2024 Intel Corporation
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/detail/cpu.cpp:2]: is it necessary to change Copyright here? +please, use the latest copyright rule

File: cpp/oneapi/dal/test/global_context.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    const uint32_t ids_count = registers.eax_;
+    if (ids_count >= 1) {
+        cpuid(0x1, registers);
+        if (registers.edx_ & (1U << 26)) { ext = detail::cpu_extension::sse2; }
+        if (registers.ecx_ & (1U << 20)) { ext = detail::cpu_extension::sse42; }
+    }
+    if (ids_count >= 7) {
+        cpuid(0x7, registers);
+        if (registers.ebx_ & (1U << 5)) { ext = detail::cpu_extension::avx2; }
+        if (registers.ebx_ & (1U << 16)) { ext = detail::cpu_extension::avx512; }
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/test/global_context.cpp:123]: is there any case when we should contain all of these instructions?
for example, if user wants to get cpu info as you did in the test?

File: cpp/oneapi/dal/global_context.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+namespace oneapi::dal {
+namespace v1 {
+
+const global_context_iface& global_context::get_global_context() {
+    [[maybe_unused]] static detail::global_context_impl init;
+    return init;
+}
+
+} // namespace v1
+} // namespace oneapi::dal
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/global_context.cpp:29]: please add a newline

File: cpp/oneapi/dal/detail/global_context_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+global_context_impl::global_context_impl() : cpu_info_(dal::detail::detect_top_cpu_extension()) {
+    using daal::services::Environment;
+    // Call to `getCpuId` changes global settings, in particular,
+    // changes default number of threads in the threading layer
+    Environment::getInstance()->getCpuId();
+}
+
+} // namespace v1
+} // namespace oneapi::dal::detail
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/detail/global_context_impl.cpp:32]: please add a newline

File: cpp/oneapi/dal/global_context.cpp (hunk 3 above)
Comment (Vika-F) [reply cpp/oneapi/dal/global_context.cpp]: done in all places where needed

File: cpp/oneapi/dal/detail/cpu_info_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+std::string cpu_info_impl::dump() const {
+    std::stringstream ss;
+    for (auto it = info_.begin(); it != info_.end(); ++it) {
+        ss << it->first << " : ";
+        print_any(it->second, ss);
+        ss << "; ";
+    }
+    std::string result;
```
Comment (keeranroth) [inline cpp/oneapi/dal/detail/cpu_info_impl.cpp:66]: This is just getting the string from the stream, right? In which case
```
  return std::move(ss).str();
```
is preferred over unpacking space separated tokens from the stream

File: cpp/oneapi/dal/detail/cpu_info_x86_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+namespace v1 {
+
+class cpu_info_x86 : public cpu_info_impl {
+public:
+    cpu_info_x86() {
+        info_["top_cpu_extension"] = detect_top_cpu_extension();
+        info_["vendor"] = (daal_check_is_intel_cpu() ? cpu_vendor::intel : cpu_vendor::amd);
+    }
+
+    cpu_info_x86(const cpu_extension cpu_extension) {
```
Comment (keeranroth) [inline cpp/oneapi/dal/detail/cpu_info_x86_impl.hpp:33]: nit: explicit

File: cpp/oneapi/dal/detail/cpu_info_arm_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+namespace v1 {
+
+class cpu_info_arm : public cpu_info_impl {
+public:
+    cpu_info_arm() {
+        info_["top_cpu_extension"] = cpu_extension::sve;
+        info_["vendor"] = cpu_vendor::arm);
+    }
+
+    cpu_info_arm(const cpu_extension cpu_extension) {
```
Comment (keeranroth) [inline cpp/oneapi/dal/detail/cpu_info_arm_impl.hpp:31]: nit: explicit

File: cpp/oneapi/dal/detail/global_context_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace oneapi::dal::detail {
+namespace v1 {
+
+class global_context_impl : public global_context_iface {
+public:
+    global_context_impl();
+
+    const cpu_info_iface &get_cpu_info() const {
+        return *dynamic_cast<const cpu_info_iface *>(&cpu_info_);
```
Comment (keeranroth) [inline cpp/oneapi/dal/detail/global_context_impl.hpp:29]: No need for rtti. The inheritance of cpu_info_iface is public in cpu_info, so you can just do:
```
const cpu_info_iface &get_cpu_info() const {
  return cpu_info_;
}
```

File: cpp/oneapi/dal/test/global_context.cpp [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/test/engine/common.hpp"
+#include "oneapi/dal/detail/global_context.hpp"
+
+namespace oneapi::dal::detail::test {
+
+#if defined(__x86_64__)
```
Comment (keeranroth) [inline cpp/oneapi/dal/test/global_context.cpp:22]: Should you be using TARGET_X86_64 here? You are later in the file

File: cpp/oneapi/dal/test/global_context.cpp [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+/// Copies the value of x86 register into the buffer of characters
+///
+/// @param[in]  reg    The x86 register value
+/// @param[out] buffer Pointer to the output character buffer
+void register_to_buffer(uint32_t reg, char* buffer) {
+    ONEDAL_ASSERT(buffer);
+    constexpr uint32_t mask = 0xFF; // lower 8 bits
+    constexpr uint32_t register_size = sizeof(uint32_t);
+    for (uint32_t i = 0; i < register_size; ++i) {
```
Comment (keeranroth) [inline cpp/oneapi/dal/test/global_context.cpp:57]: Looks like little-endian write of bytes, so you don't need the copy like this? The uint32_t are little endian on x86, so you can just write out the value. Assuming 8-bit char, of course
```
ONEDAL_ASSERT(buffer);
*(uint32_t*)buffer = reg;
```

File: cpp/oneapi/dal/test/global_context.cpp [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+/// Copies the values of x86 EBX, EDX, ECX (in that order) registers that contain the vendor ID
+/// into the buffer of characters after the call of CPUID with EAX == 0
+///
+/// @param[in]  registers   x86 CPU registers with the CPUID results
+/// @param[out] buffer      Pointer to the output character buffer
+void registers_to_vendor_id(const cpuid_registers& registers, char* buffer) {
+    ONEDAL_ASSERT(buffer);
+    register_to_buffer(registers.ebx_, buffer);
+    register_to_buffer(registers.edx_, buffer + 4);
```
Comment (keeranroth) [inline cpp/oneapi/dal/test/global_context.cpp:71]: sizeof(uint32_t) instead of 4, and 2 * sizeof(uint32_t) for the + 8, just to be explicit

**Comment by Vika-F** [comment]
@keeranroth Than you for the thorough review. I find all your comments valid; now in the process of implementing the suggestions.
I'll probably request one more review iteration as soon as the fixes are ready.

**Review by keeranroth — APPROVED** [review]
Looks good to me. Thanks :)

**Review by Alexandr-Solovev — APPROVED** [review]

**Pull request merged**